Sprint 2/step 1

Nikba-Creative-Studio opened the pull request:
Проект. Перенос состояния в Context

1. Использование Context в конструкторе бургеров
Данные перенесены в Context (services/constructorContext.js)
2. Набор ограничений внутри конструктора бургера
- только один ингредиент с типом bun
-  данные с типом bun ренрерется 2 раза ( верх и низ )
-  для проверки вычислений итоговой стоимости реализованы функционал для удаления ингредиента из конструктора
3. Подсчёт итоговой стоимости бургера
Счетается общая цена ингредиентов из массива конструктора
4. Возможность создавать заказ
- реализован функционал для отправки заказа на бэкэнд через post запрос
- полученные данные отправляются в OrderDetails
5. Проверка типизации
- добавлено для OrderDetails
6. Дополнительно
- Функционал модальных окон выведен в App
- Разработан функционал  ScroolTo для табов ингредиентов
- Разработан функционал для показа количество определенного ингредиента в конструкторе

Nikba-Creative-Studio committed:
Рефакторинг Модального Окна

Поднял фунционал модальных окон на уровень App

Nikba-Creative-Studio committed:
Использование Context в конструкторе бургеров

Nikba-Creative-Studio committed:
Функция создать заказ

Nikba-Creative-Studio committed:
Отправка данных заказа в OrderDetails

Nikba-Creative-Studio committed:
Рефакторинг BurgerIngredients

1. Оптимизирован код по принципу DRY
2. Добавлен фунционал для скрулла
3. Компонент Indredient реагирует на кол. ингредиентов в конструкторе

Nikba-Creative-Studio committed:
Update burger-constructor.js

irinamozes commented:
Проект очень неплохой.

Динамически рассчитывается стоимость бургера.

Передаётся контекст в подписанные на него компоненты.

Что нужно подкорректировать.

1. При рендере списков методом map, в атрибуте key нужно обязательно задавать идентификатор элемента, если у элемента есть уникальный идентификатор, а не индекс массива, по которому идёт рендер.  Нужно, чтобы реакт идентифицировал элементы по их _id, а не по индексу массива. Об этом можно прочитать в Ваших лекциях с тренажёрами (о списках и ключах) , которые относятся к 1-му этапу 1-го спринта.
Поэтому при рендере списка ингредиентов (в левой части главной страницы) в map нужно использовать идентификатор ингредиента (Это Вы делаете правильно).
Однако, при рендере массива слоёв в бургере нужно использовать индекс массива (на этом этапе), так как в бургере могут быть одинаковые ингредиенты и идентификатор элемента списка может быть не уникальным (если бы в Вашем экспериментальном бургере были одинаковые слои, реакт бы выдал предупреждение в консоли о неуникальности атрибута key). Это будет очень важно для следующпего этапа. То есть при рендере массива слоёв в бургере должно
быть key={index} (на этом этапе, на следующем этапе должен будет создаваться уникальный случайный ключ для каждого слоя).

2. При закрытии модального окна (с любым содержимом) по оверлею и Esc приложение падает с ошибкой в консоли (см. снимок):
![Снимок экрана в 2022-02-02 15-00-14](https://user-images.githubusercontent.com/19558891/152139680-1c4a476a-f1e3-4943-8700-812e0a23ed98.png)


Что можно лучше.

1. Лучше как начальное значение параметра useState задавать не пустой объект ({}), а null. Хотя бы потому, что null легче проверять на ложное значение, так как js рассматривает null как false. Пустой объект ({}) false не является.
Прочитайте об этом статьи: https://russianblogs.com/article/8801745392/, https://qna.habr.com/q/928113.

Nikba-Creative-Studio committed:
Исправление багов

1. устранение ошибки при включении модального окна. Задаем null в useState вместо пустого объекта
2. Уникальный id для key в конструкторе

Nikba-Creative-Studio committed:
Update ingredient-details.js

irinamozes commented:
Проект отличный!

Не забудьте, пожалуйста, что у булок в собранном бургере должны быть иконки замков, а не корзинок.

Рекомендация: во всех последующих заданиях до конца курсов не нужно использовать хуки useReducer, по-возможности хук useState, и особенно контекст, так как их полностью заменит сама технология Redux и хук useSelector, который, заменяя контекст, уменьшает количество лишних перерисовок компонент.

Задание принимается.

Желаю дальнейших успехов в учёбе и удачи!
